File: app/src/main/java/com/example/myaac/data/remote/ArasaacService.kt
package com.example.myaac.data.remote

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

data class ArasaacPictogram(
    val _id: Int,
    val keywords: List<ArasaacKeyword> = emptyList()
)

data class ArasaacKeyword(
    val keyword: String
)

open class ArasaacService : SymbolService {
    private val BASE_URL = "https://api.arasaac.org/v1/"

    private val api: ArasaacApi by lazy {
        retrofit2.Retrofit.Builder()
            .baseUrl(BASE_URL)
            .addConverterFactory(retrofit2.converter.gson.GsonConverterFactory.create())
            .build()
            .create(ArasaacApi::class.java)
    }
    
    // Manual overrides for common ambiguous words
    private val SYMBOL_OVERRIDES = mapOf(
        "can" to 26521, // Verb "can" / "able to"
        "will" to 26497, // Future tense arrow
        "may" to 22350,  // "Maybe" / possibility
        "like" to 2267,  // Thumbs up / like
        "kind" to 34166, // Kind/nice person
        "saw" to 5858,   // Verb "see" (past) instead of tool
        "well" to 34166, // Good/Well
        "mine" to 33744, // Possessive
        // Pronouns
        "i" to 6632,
        "you" to 6625,
        "we" to 7185,
        "they" to 7032,
        "he" to 6480,
        "she" to 7028,
        // Core Verbs
        "want" to 5441,
        "need" to 37160,
        "feel" to 30197,
        "have" to 32761,
        "go" to 8142,
        // Questions
        "what" to 22620,
        "where" to 7764,
        "when" to 32874,
        "why" to 36719,
        // Be verbs if needed
        "am" to 26521, // Reuse 'can/ability' or find better
        "is" to 26521,
        "are" to 26521
    )

    open suspend fun searchPictograms(query: String, locale: String = "en"): List<ArasaacPictogram> {
        // Check overrides first (only for English single words for now)
        if (locale == "en") {
            val lowerQuery = query.lowercase().trim()
            val overrideId = SYMBOL_OVERRIDES[lowerQuery]
            if (overrideId != null) {
                return listOf(ArasaacPictogram(overrideId, listOf(ArasaacKeyword(query))))
            }
        }
    
        return withContext(Dispatchers.IO) {
            api.searchPictograms(locale, query)
        }
    }
    
    open fun getImageUrl(id: Int): String {
        return "https://static.arasaac.org/pictograms/$id/${id}_300.png"
    }

    override suspend fun search(query: String, language: String): List<SymbolResult> {
        val pictograms = searchPictograms(query, language)
        return pictograms.map { 
            val label = it.keywords.firstOrNull()?.keyword ?: query
            SymbolResult(
                url = getImageUrl(it._id),
                label = label
            )
        }
    }
}

File: web/src/services/arasaacService.ts
// Port of data/remote/ArasaacService.kt (+ ArasaacApi.kt) — near-mechanical translation.
// REST: GET https://api.arasaac.org/v1/pictograms/{locale}/search/{query}

import type { SymbolResult, SymbolService } from './symbolService';

export interface ArasaacKeyword {
  keyword: string;
}

export interface ArasaacPictogram {
  _id: number;
  keywords: ArasaacKeyword[];
}

const BASE_URL = 'https://api.arasaac.org/v1/';

/**
 * Manual overrides for common ambiguous words.
 * Ported verbatim from ArasaacService.kt's SYMBOL_OVERRIDES map.
 */
const SYMBOL_OVERRIDES: Record<string, number> = {
  can: 26521, // Verb "can" / "able to"
  will: 26497, // Future tense arrow
  may: 22350, // "Maybe" / possibility
  like: 2267, // Thumbs up / like
  kind: 34166, // Kind/nice person
  saw: 5858, // Verb "see" (past) instead of tool
  well: 34166, // Good/Well
  mine: 33744, // Possessive
  // Pronouns
  i: 6632,
  you: 6625,
  we: 7185,
  they: 7032,
  he: 6480,
  she: 7028,
  // Core Verbs
  want: 5441,
  need: 37160,
  feel: 30197,
  have: 32761,
  go: 8142,
  // Questions
  what: 22620,
  where: 7764,
  when: 32874,
  why: 36719,
  // Be verbs if needed
  am: 26521, // Reuse 'can/ability' or find better
  is: 26521,
  are: 26521,
};

export class ArasaacService implements SymbolService {
  async searchPictograms(query: string, locale: string = 'en'): Promise<ArasaacPictogram[]> {
    // Check overrides first (only for English single words for now)
    if (locale === 'en') {
      const lowerQuery = query.toLowerCase().trim();
      const overrideId = SYMBOL_OVERRIDES[lowerQuery];
      if (overrideId != null) {
        return [{ _id: overrideId, keywords: [{ keyword: query }] }];
      }
    }

    const url = `${BASE_URL}pictograms/${encodeURIComponent(locale)}/search/${encodeURIComponent(query)}`;
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`ArasaacService: HTTP ${response.status}`);
    }
    return (await response.json()) as ArasaacPictogram[];
  }

  getImageUrl(id: number): string {
    return `https://static.arasaac.org/pictograms/${id}/${id}_300.png`;
  }

  async search(query: string, language: string): Promise<SymbolResult[]> {
    const pictograms = await this.searchPictograms(query, language);
    return pictograms.map((p) => {
      const label = p.keywords[0]?.keyword ?? query;
      return {
        url: this.getImageUrl(p._id),
        label,
      };
    });
  }
}
